File: frontend/services/administration.ts
const apiBaseUrl = import.meta.env.VITE_API_BASE_URL ?? "http://localhost:8000";

export type Permission = { code: string; description: string | null };
export type Role = { id: string; name: string; description: string | null; is_active: boolean; is_system_managed: boolean; permission_codes: string[] };
export type Location = { id: string; name: string; is_active: boolean; branch_id?: string | null };
export type CompanyUser = {
  id: string; name: string; email: string; mobile_number: string; account_state: string; email_verified: boolean; mobile_verified: boolean;
  company_access_active: boolean; role_ids: string[]; role_names: string[]; default_branch_id: string | null; default_warehouse_id: string | null; has_active_session: boolean;
};
export type AuditEvent = { id: string; action: string; actor_user_id: string | null; target_user_id: string | null; details: string | null; created_at: string };
export type Bootstrap = { users: CompanyUser[]; roles: Role[]; permissions: Permission[]; branches: Location[]; warehouses: Location[]; audit_events: AuditEvent[]; effective_permissions: string[] };
export type UserPayload = { name: string; email?: string; mobile_number?: string; role_ids: string[]; default_branch_id: string | null; default_warehouse_id: string | null };
export type RolePayload = { name: string; description: string | null; permission_codes: string[]; is_active: boolean };

async function request<T>(path: string, companyId: string, method = "GET", body?: object): Promise<T> {
  const response = await fetch(`${apiBaseUrl}${path}`, { method, credentials: "include", headers: { "X-Company-ID": companyId, ...(body ? { "Content-Type": "application/json" } : {}) }, body: body ? JSON.stringify(body) : undefined });
  const payload = await response.json().catch(() => ({}));
  if (!response.ok) throw new Error(payload.detail ?? "Unable to complete this administration action.");
  return payload as T;
}

export const administrationApi = {
  bootstrap: (companyId: string) => request<Bootstrap>("/administration/bootstrap", companyId),
  invite: (companyId: string, payload: UserPayload) => request<CompanyUser>("/administration/users", companyId, "POST", payload),
  updateUser: (companyId: string, userId: string, payload: UserPayload) => request<CompanyUser>(`/administration/users/${userId}`, companyId, "PUT", payload),
  setUserStatus: (companyId: string, userId: string, is_active: boolean) => request<CompanyUser>(`/administration/users/${userId}/status`, companyId, "PATCH", { is_active }),
  forceLogout: (companyId: string, userId: string) => request<{ message: string }>(`/administration/users/${userId}/force-logout`, companyId, "POST", {}),
  reset: (companyId: string, userId: string) => request<{ message: string }>(`/administration/users/${userId}/reset`, companyId, "POST", {}),
  createRole: (companyId: string, payload: RolePayload) => request<Role>("/administration/roles", companyId, "POST", payload),
  updateRole: (companyId: string, roleId: string, payload: RolePayload) => request<Role>(`/administration/roles/${roleId}`, companyId, "PUT", payload),
  cloneRole: (companyId: string, roleId: string, payload: RolePayload) => request<Role>(`/administration/roles/${roleId}/clone`, companyId, "POST", payload),
};

File: backend/app/schemas/administration.py
"""Company-scoped user, role, permission, and security administration contracts."""

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class PermissionResponse(BaseModel):
    code: str
    description: str | None


class RoleResponse(BaseModel):
    id: UUID
    name: str
    description: str | None
    is_active: bool
    is_system_managed: bool
    permission_codes: list[str]


class RoleInput(BaseModel):
    name: str = Field(min_length=2, max_length=100)
    description: str | None = Field(default=None, max_length=2000)
    permission_codes: list[str] = Field(default_factory=list)
    is_active: bool = True


class CompanyUserResponse(BaseModel):
    id: UUID
    name: str
    email: str
    mobile_number: str
    account_state: str
    email_verified: bool
    mobile_verified: bool
    company_access_active: bool
    role_ids: list[UUID]
    role_names: list[str]
    default_branch_id: UUID | None
    default_warehouse_id: UUID | None
    has_active_session: bool


class UserInviteRequest(BaseModel):
    name: str = Field(min_length=2, max_length=255)
    email: str = Field(min_length=3, max_length=320)
    mobile_number: str = Field(min_length=6, max_length=20)
    role_ids: list[UUID] = Field(min_length=1)
    default_branch_id: UUID | None = None
    default_warehouse_id: UUID | None = None


class UserUpdateRequest(BaseModel):
    name: str = Field(min_length=2, max_length=255)
    role_ids: list[UUID] = Field(default_factory=list)
    default_branch_id: UUID | None = None
    default_warehouse_id: UUID | None = None


class ActiveStatusRequest(BaseModel):
    is_active: bool


class AuditEventResponse(BaseModel):
    id: UUID
    action: str
    actor_user_id: UUID | None
    target_user_id: UUID | None
    details: str | None
    created_at: datetime


class AdministrationBootstrapResponse(BaseModel):
    users: list[CompanyUserResponse]
    roles: list[RoleResponse]
    permissions: list[PermissionResponse]
    branches: list[dict[str, object]]
    warehouses: list[dict[str, object]]
    audit_events: list[AuditEventResponse]
    effective_permissions: list[str]
